joining/src/main/scala/akka/cluster/bootstrap/dns/HeadlessServiceDnsBootstrap.scala

```diff
@@ -0,0 +1,278 @@
+ * this will mean it joins the "right" cluster.
+ *
+ * CAVEATS:
+ * There is a slight timing issue, that may theoretically appear in this bootstrap process.
+ * FIXME explain the races
+ */
+// also known as the "Baron von Bootstrappen"
+@InternalApi
+final class HeadlessServiceDnsBootstrap(settings: ClusterBootstrapSettings)
+    extends Actor
+    with ActorLogging
+    with Timers {
+
+  import HeadlessServiceDnsBootstrap.Protocol._
+  import HeadlessServiceDnsBootstrap._
+  import context.dispatcher
+
+  private val dns = IO(Dns)(context.system)
+  private val cluster = Cluster(context.system)
+
+  private val TimerKeyResolveDNS = "resolve-dns-key"
+
+  private var lastContactsObservation: DnsServiceContactsObservation =
+    DnsServiceContactsObservation(Long.MaxValue, Nil)
+
+  /** Awaiting initial signal to start the bootstrap process */
+  override def receive: Receive = {
+    case InitiateBootstraping ⇒
+      val serviceName = settings.effectiveServiceName(context.system)
+
+      log.info("Locating service members, via DNS lookup: {}", serviceName)
+      resolve(serviceName, settings.dnsResolveTimeout).pipeTo(self)
+
+      context become bootstraping(serviceName, sender())
+  }
+
+  /** In process of searching for seed-nodes */
+  def bootstraping(serviceName: String, replyTo: ActorRef): Receive = {
+    case Internal.AttemptResolve(name, timeout) ⇒
+      resolve(name, timeout).pipeTo(self)
+
+    case DnsResolvedTo(name, contactPoints) ⇒
+      onContactPointsResolved(name, contactPoints)
+
+    case ex: Failure ⇒
+      log.warning("Resolve attempt failed! Cause: {}", ex.cause)
+      scheduleNextResolve(serviceName, settings.dnsResolveTimeout)
+
+    case ObtainedHttpSeedNodesObservation(infoFromAddress, observedSeedNodes) ⇒
+      log.info("Contact point [{}] returned [{}] seed-nodes [{}], initiating cluster joining...", infoFromAddress,
+        observedSeedNodes.size, observedSeedNodes)
+
+      replyTo ! "JOINING SEED NODES!" // FIXME nicer typed message
+
+      val seedNodesList = observedSeedNodes.toList
+      cluster.joinSeedNodes(seedNodesList)
+
+      // once we issued a join bootstraping is completed
+      context.stop(self)
+
+    case NoSeedNodesObtainedWithinDeadline(contactPoint) ⇒
+      log.info(
+          "Contact point [{}] exceeded stable margin with no seed-nodes in sight. " +
+          "Considering weather this node is allowed to JOIN itself to initiate a new cluster.",
+          contactPoint) // TODO debug
+
+      onNoSeedNodesObtainedWithinStableDeadline(contactPoint)
+  }
+
+  private def onContactPointsResolved(serviceName: String, contactPoints: immutable.Seq[String]): Unit = {
+    val newObservation = DnsServiceContactsObservation(timeNow(), contactPoints.toList)
+    lastContactsObservation = lastContactsObservation.sameOrChanged(newObservation)
+
+    if (contactPoints.size < settings.requiredContactPointsNr)
```

Comment:
as discussed in Slack we should perhaps rename it to `initial-nr-of-contact-points` and it must be exactly that number of contact points before self-joining.

I'm thinking about if it's needed to wait here before starting probing? It should be enough with one other contact point and if that returns some seed-nodes we can join that?

For example you might want to set `initial-nr-of-contact-points` to 5, becuase you know that you are starting 5 nodes initially, but for later restarts it should be possible to join with less than 5 nodes. E.g. 3 of 5 nodes crash and 2 nodes are started again it's better that those can join than having to wait for a 3rd.